### PR TITLE
feat(service-worker): visibility showNotification.

### PR DESCRIPTION
**Pull request by kevinlllR**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A
Currently when a push notification is received, the default showNotification interface is displayed.

## What is the new behavior?
That the showNotification interface is displayed only if a visibility flag is sent and the clients are hidden, otherwise the current behavior is followed.
This allows you to customize the notifications that a client receives when it is visible and not show double.
## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by kevinlllR** [comment]
> @googlebot I signed it!

**Review by gkalpak — CHANGES_REQUESTED** [review]
Thx for taking the time to submit the PR, @kevinlllR.

There are a couple of things that need to be fixed, before we can consider the PR for merging (e.g. add tests, rebase on master and get CI passing, make the commit message more descriptive, etc.)

But first, I want to better understand what is the feature that you are adding. Looking at the code and the PR description, it is not 100% clear to me. Can you, please, add some more info?

##
Also, if I understand the intention correctly, this would be adding a proprietary API/requirement for notifications handled by Angular apps (to have a `visible` property). We generally try to avoid that, since it makes it harder for people to integrate push notifications into apps. (I know we have the custom `.notifaction` property, but we are trying to keep them at a minimum and add new ones only if necessary.)

Couldn't your usecase be handled on the application level? I.e. all applications would be notified, but only the visible ones would choose to react to the notification. You could use something like [Document#visibilityState][1] to track the visibility state of each page.

[1]: https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState

File: packages/service-worker/worker/src/service-worker.d.ts [hunk 1]
```diff
@@ -37,15 +37,15 @@ declare class Client {
 interface Clients {
   claim(): Promise<any>;
   get(id: string): Promise<Client>;
-  matchAll(options?: ClientMatchOptions): Promise<Array<Client>>;
+  matchAll(options?: ClientMatchOptions): Promise<Array<WindowClient>>;
```
[the remark this answers is not in the document]
Comment (kevinlllR) [reply packages/service-worker/worker/src/service-worker.d.ts]: Thanks, I will correct the error.

[the remark this answers is not in the document]
Comment (kevinlllR) [reply packages/service-worker/worker/src/service-worker.d.ts]: When I run the tests, they all pass, but in the end there is an error with yarn, should I open a separate thread?

```
INFO: Elapsed time: 0.485s, Critical Path: 0.03s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
error Command failed with exit code 4.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/service-worker/worker/src/service-worker.d.ts]: No idea what this is about. Yes, please, open a new issue.
Does it happen when you run the tests on master too or only after the PR changes?

[the remark this answers is not in the document]
Comment (kevinlllR) [reply packages/service-worker/worker/src/service-worker.d.ts]: > Thx for taking the time to submit the PR, @kevinlllR.
> 
> There are a couple of things that need to be fixed, before we can consider the PR for merging (e.g. add tests, rebase on master and get CI passing, make the commit message more descriptive, etc.)
> 
> But first, I want to better understand what is the feature that you are adding. Looking at the code and the PR description, it is not 100% clear to me. Can you, please, add some more info?
> 
> Also, if I understand the intention correctly, this would be adding a proprietary API/requirement for notifications handled by Angular apps (to have a `visible` property). We generally try to avoid that, since it makes it harder for people to integrate push notifications into apps. (I know we have the custom `.notifaction` property, but we are trying to keep them at a minimum and add new ones only if necessary.)
> 
> Couldn't your usecase be handled on the application level? I.e. all applications would be notified, but only the visible ones would choose to react to the notification. You could use something like [Document#visibilityState](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState) to track the visibility state of each page.



> Thx for taking the time to submit the PR, @kevinlllR.
> 
> There are a couple of things that need to be fixed, before we can consider the PR for merging (e.g. add tests, rebase on master and get CI passing, make the commit message more descriptive, etc.)
> 
> But first, I want to better understand what is the feature that you are adding. Looking at the code and the PR description, it is not 100% clear to me. Can you, please, add some more info?
> 
> Also, if I understand the intention correctly, this would be adding a proprietary API/requirement for notifications handled by Angular apps (to have a `visible` property). We generally try to avoid that, since it makes it harder for people to integrate push notifications into apps. (I know we have the custom `.notifaction` property, but we are trying to keep them at a minimum and add new ones only if necessary.)
> 
> Couldn't your usecase be handled on the application level? I.e. all applications would be notified, but only the visible ones would choose to react to the notification. You could use something like [Document#visibilityState](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState) to track the visibility state of each page.


Hello, regarding the use cases, I want to reduce the dissatisfaction of receiving notifications constantly (in the case of advertising, it is a bit heavy), I have a lot of periodically scheduled notifications, there are times that a person can reach 20 or 30 notifications in less than a minute, with document.ShowNotification, I cannot intercept that users receive the default notifications of showNotification, and at the same time mine, when notifications are not scheduled, with document.ShowNotifaction I could intercept every time a message is received by a user, and based on it send the push notification, but the visibility status is something that changes constantly, and if a push notification was sent, I see no way to cancel it, which would cause a user to receive equal double notification (take into account that processing the request validate and send the notification has a delay), in addition to document.Shownotificat I can only control the visibility status of a tab, but if a user has 1 tab open and 4 minimized, I would have 4 push notifications, which by default will show me 4 showNotification, fix these details I guess there are options, but they don't occur to me and if so, it would be very cumbersome.

In summary, putting a visible flag to the data that is sent in the notification, would allow to avoid these annoying ones, in addition to reducing the queries to the server in case of multiple tabs.

[the remark this answers is not in the document]
Comment (kevinlllR) [reply packages/service-worker/worker/src/service-worker.d.ts]: > No idea what this is about. Yes, please, open a new issue.
> Does it happen when you run the tests on master too or only after the PR changes?

It happens to me both in master and in my branch, delete cache from yarn, from bazel and the same, I cloned the angular repository and the same thing happens, I don't have other devices to test.
[issue](https://github.com/angular/angular/issues/32507)

[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/service-worker/worker/src/service-worker.d.ts]: Sorry, @kevinlllR, I am still not sure I understand your usecase. Could you try to explain it again?
It sounds like it is something too specific, so likely not a good fit to incorporate into `@angular/service-worker`.

**Comment by gkalpak** [comment]
Closing due to lack of response.
This is likely too specific a usecase to be suitable for `@angular/service-worker`, but feel free to open a new issue to discuss.
